### PR TITLE
Simple handling of mid-call output changes

### PR DESCRIPTION
**Pull request by SimonBrandner**
This converts the `CallMediaHandler` into a `MediaDeviceHandler` which is an `EventEmitter`. 

We should add some UI to the `CallView` for this but this should work for now

### CODE REVIEW

File: src/MediaDeviceHandler.ts [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+        }
+    }
+
+    static loadDevices() {
+        const audioDeviceId = SettingsStore.getValue("webrtc_audioinput");
+        const videoDeviceId = SettingsStore.getValue("webrtc_videoinput");
+
+        setMatrixCallAudioInput(audioDeviceId);
+        setMatrixCallVideoInput(videoDeviceId);
+    }
```
Comment (t3chguy) [inline src/MediaDeviceHandler.ts:85]: what's the visibility of these methods?

File: src/MediaDeviceHandler.ts [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+                audioOutput: audioOutput,
+                audioInput: audioInput,
+                videoInput: videoInput,
+            };
+        } catch (error) {
+            console.log('Unable to refresh WebRTC Devices: ', error);
+        }
+    }
+
+    static loadDevices() {
```
Comment (t3chguy) [inline src/MediaDeviceHandler.ts:79]:
```suggestion
    static loadDevices(): void {
```

File: src/MediaDeviceHandler.ts [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+        SettingsStore.setValue("webrtc_audioinput", null, SettingLevel.DEVICE, deviceId);
+        setMatrixCallAudioInput(deviceId);
+        this.emit(MediaDeviceHandlerEvent.AudioInputChanged, deviceId);
+    }
+
+    public setVideoInput(deviceId: string) {
+        SettingsStore.setValue("webrtc_videoinput", null, SettingLevel.DEVICE, deviceId);
+        setMatrixCallVideoInput(deviceId);
+        this.emit(MediaDeviceHandlerEvent.VideoInputChanged, deviceId);
+    }
```
Comment (t3chguy) [inline src/MediaDeviceHandler.ts:102]:
```suggestion
    public setAudioOutput(deviceId: string): void {
        SettingsStore.setValue("webrtc_audiooutput", null, SettingLevel.DEVICE, deviceId);
        this.emit(MediaDeviceHandlerEvent.AudioOutputChanged, deviceId);
    }

    public setAudioInput(deviceId: string): void {
        SettingsStore.setValue("webrtc_audioinput", null, SettingLevel.DEVICE, deviceId);
        setMatrixCallAudioInput(deviceId);
        this.emit(MediaDeviceHandlerEvent.AudioInputChanged, deviceId);
    }

    public setVideoInput(deviceId: string): void {
        SettingsStore.setValue("webrtc_videoinput", null, SettingLevel.DEVICE, deviceId);
        setMatrixCallVideoInput(deviceId);
        this.emit(MediaDeviceHandlerEvent.VideoInputChanged, deviceId);
    }
```

File: src/MediaDeviceHandler.ts [hunk 4]
```diff
@@ -68,18 +84,21 @@ export default class MediaDeviceHandler {
...
         SettingsStore.setValue("webrtc_audioinput", null, SettingLevel.DEVICE, deviceId);
         setMatrixCallAudioInput(deviceId);
+        this.emit(MediaDeviceHandlerEvent.AudioInputChanged, deviceId);
     }
 
-    static setVideoInput(deviceId: string) {
+    public setVideoInput(deviceId: string) {
         SettingsStore.setValue("webrtc_videoinput", null, SettingLevel.DEVICE, deviceId);
         setMatrixCallVideoInput(deviceId);
+        this.emit(MediaDeviceHandlerEvent.VideoInputChanged, deviceId);
```
Comment (t3chguy) [inline src/MediaDeviceHandler.ts]: Will these work mid-call at this point or not?

Comment (SimonBrandner) [reply src/MediaDeviceHandler.ts]: No, that would require, js-sdk changes. I hope to do that at some point too, though

Comment (t3chguy) [reply src/MediaDeviceHandler.ts]: That's what I thought, can we get comments stating that at the very least? It seems weird it emits an event that we know will leave the app in a split-brained state if it was actually listening on it

File: src/MediaDeviceHandler.ts [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+                    case 'audioinput': audioInput.push(device); break;
+                    case 'videoinput': videoInput.push(device); break;
+                }
+            });
+
+            return {
+                audioOutput: audioOutput,
+                audioInput: audioInput,
+                videoInput: videoInput,
+            };
```
Comment (t3chguy) [inline src/MediaDeviceHandler.ts:71]:
```suggestion
            return { audioOutput, audioInput, videoInput };
```

Just making sure you're aware of this short-hand, can be cleaner

File: src/MediaDeviceHandler.ts [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+        } catch (error) {
+            console.log('Unable to refresh WebRTC Devices: ', error);
```
Comment (t3chguy) [inline src/MediaDeviceHandler.ts:73]: this should be warn at least

File: src/MediaDeviceHandler.ts (hunk 5 above)
Comment (SimonBrandner) [reply src/MediaDeviceHandler.ts]: I am semi-aware of it, I seem to always forget about it :D

File: src/components/views/voip/AudioFeed.tsx [hunk 7]
```diff
@@ -27,19 +27,25 @@ export default class AudioFeed extends React.Component<IProps> {
...
+        MediaDeviceHandler.instance.removeListener(
+            MediaDeviceHandlerEvent.AudioOutputChanged,
+            this.onAudioOutputChanged,
+        );
         this.props.feed.removeListener(CallFeedEvent.NewStream, this.onNewStream);
         this.stopMedia();
     }
 
-    private playMedia() {
+    private onAudioOutputChanged = (audioOutput: string) => {
```
Comment (t3chguy) [inline src/components/views/voip/AudioFeed.tsx:47]: Hmm, seems like setSinkId now only gets called if the audio output changed during this session

**Review by t3chguy — APPROVED** [review]
Looks sane to me, thanks

**Pull request merged**